overlord/configstate/configcore/certs.go

```diff
@@ -0,0 +1,112 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+)
+
+func handleCertConfiguration(tr config.Conf) error {
+	// This handles the "snap revert core" case:
+	// We need to go over each pem cert on disk and check if there is
+	// a matching config entry - if not->delete the cert
+	//
+	// XXX: remove this code once we have a general way to handle
+	//      "snap revert" and config updates
+	storeCerts, err := filepath.Glob(filepath.Join(dirs.SnapdStoreSSLCertsDir, "*.pem"))
+	if err != nil {
+		return fmt.Errorf("cannot get exiting store certs: %v", err)
+	}
+	for _, storeCertPath := range storeCerts {
+		optionName := strings.TrimSuffix(filepath.Base(storeCertPath), ".pem")
+		v, err := coreCfg(tr, "certs."+optionName)
+		if err != nil {
+			return err
+		}
+		if v == "" {
+			if err := os.Remove(storeCertPath); err != nil {
+				return err
+			}
+		}
+	}
+
+	// add/remove regular (non revert) changes
+	for _, name := range tr.Changes() {
+		if !strings.HasPrefix(name, "core.certs.") {
+			continue
+		}
+
+		nameWithoutSnap := strings.SplitN(name, ".", 2)[1]
+		cert, err := coreCfg(tr, nameWithoutSnap)
+		if err != nil {
+			return fmt.Errorf("internal error: cannot get data for %s: %v", nameWithoutSnap, err)
+		}
+		optionName := strings.SplitN(name, ".", 3)[2]
+		certPath := filepath.Join(dirs.SnapdStoreSSLCertsDir, optionName+".pem")
+		switch cert {
+		case "":
+			// remove
+			if err := os.Remove(certPath); err != nil && !os.IsNotExist(err) {
+				return fmt.Errorf("cannot remove cert: %v", err)
+			}
+		default:
+			if err := os.MkdirAll(dirs.SnapdStoreSSLCertsDir, 0755); err != nil {
+				return fmt.Errorf("cannot create store ssl certs dir: %v", err)
+			}
+			if err := ioutil.WriteFile(certPath, []byte(cert), 0644); err != nil {
+				return fmt.Errorf("cannot write extra cert: %v", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func validateCertSettings(tr config.Conf) error {
+	for _, name := range tr.Changes() {
+		if !strings.HasPrefix(name, "core.certs.") {
+			continue
+		}
+
+		nameWithoutSnap := strings.SplitN(name, ".", 2)[1]
+		cert, err := coreCfg(tr, nameWithoutSnap)
+		if err != nil {
+			return fmt.Errorf("internal error: cannot get data for %s: %v", nameWithoutSnap, err)
+		}
+		if cert != "" {
+			optionName := strings.SplitN(name, ".", 3)[2]
+			block, rest := pem.Decode([]byte(cert))
+			if block == nil || block.Type != "CERTIFICATE" || len(rest) > 0 {
```

Comment:
the certs could be a chain I think in which case we need to check rest and not disallow it